prov/mrail/src/mrail_domain.c

```diff
@@ -0,0 +1,176 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+static int mrail_domain_close(fid_t fid)
+{
+	struct mrail_domain *mrail_domain =
+		container_of(fid, struct mrail_domain, util_domain.domain_fid.fid);
+	int ret, retv = 0;
+
+	if (mrail_domain->domains) {
+		ret = mrail_close_fids((struct fid **)mrail_domain->domains,
+				       mrail_domain->num_domains);
+		if (ret)
+			retv = ret;
+		free(mrail_domain->domains);
+	}
+
+	ret = ofi_domain_close(&mrail_domain->util_domain);
+	if (ret)
+		retv = ret;
+
+	free(mrail_domain);
+	return retv;
+}
+
+//static int mrail_mr_close(fid_t fid)
+//{
+//
+//	return -FI_ENOSYS;
+//}
+
+static int mrail_mr_reg(struct fid *domain_fid, const void *buf, size_t len,
+			 uint64_t access, uint64_t offset, uint64_t requested_key,
+			 uint64_t flags, struct fid_mr **mr, void *context)
+{
+
+	return -FI_ENOSYS;
+}
+
+//static struct fi_ops mrail_mr_ops = {
+//	.size = sizeof(struct fi_ops),
+//	.close = mrail_mr_close,
+//	.bind = fi_no_bind,
+//	.control = fi_no_control,
+//	.ops_open = fi_no_ops_open,
+//};
+
+static struct fi_ops_mr mrail_domain_mr_ops = {
+	.size = sizeof(struct fi_ops_mr),
+	.reg = mrail_mr_reg,
+	.regv = fi_no_mr_regv,
+	.regattr = fi_no_mr_regattr,
+};
+
+static struct fi_ops mrail_domain_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = mrail_domain_close,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_domain mrail_domain_ops = {
+	.size = sizeof(struct fi_ops_domain),
+	.av_open = fi_no_av_open,
+	.cq_open = mrail_cq_open,
+	.endpoint = mrail_ep_open,
+	.scalable_ep = fi_no_scalable_ep,
+	.cntr_open = fi_no_cntr_open,
+	.poll_open = fi_no_poll_open,
+	.stx_ctx = fi_no_stx_context,
+	.srx_ctx = fi_no_srx_context,
+	.query_atomic = fi_no_query_atomic,
+};
+
+int mrail_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		      struct fid_domain **domain, void *context)
+{
+	struct mrail_fabric *mrail_fabric =
+		container_of(fabric, struct mrail_fabric, util_fabric.fabric_fid);
+	struct mrail_domain *mrail_domain;
+	struct fi_info *fi, *core_info;
+	size_t i;
+	int ret;
+
+	mrail_domain = calloc(1, sizeof(*mrail_domain));
+	if (!mrail_domain)
+		return -FI_ENOMEM;
+
+	ret = ofi_domain_init(fabric, info, &mrail_domain->util_domain, context);
+	if (ret) {
+		free(mrail_domain);
+		return ret;
+	}
+
+	if (!mrail_fabric->fabrics) {
+		mrail_info_foreach(i, fi, info)
+			mrail_fabric->num_fabrics++;
+
+		mrail_fabric->fabrics = calloc(mrail_fabric->num_fabrics,
+					       sizeof(*mrail_fabric->fabrics));
+		if (!mrail_fabric->fabrics) {
+			ret = -FI_ENOMEM;
+			goto err1;
+		}
+	}
+
+	mrail_domain->num_domains = mrail_fabric->num_fabrics;
+
+	if (!(mrail_domain->domains = calloc(mrail_domain->num_domains,
+					     sizeof(*mrail_domain->domains)))) {
+		ret = -FI_ENOMEM;
+		goto err1;
+	}
+
+	mrail_info_foreach(i, fi, info) {
```

Comment:
I had a fairly lengthy discussion with Arun on this.  A multi-rail fabric is a virtual 'fabric' defined as a set of addresses, where each address represents a single 'rail'.  You can think of a rail as a core fabric + domain + endpoint.

A multi-rail fabric and domain are both virtual concepts that are basically containers.  When the mrail fabric is opened, we'll basically know what underlying domains will be associated with it.  There's a 1:many relationship between an mrail fabric and underlying domains.

The domains are not stored with the fabric though, because we don't know how many mrail domains the user will open.  Each new mrail domain opens a new set of core domains.  If 2 rails belong to the same core domain, we only open the core domain once per mrail domain.

Reply:
@shefty Thanks. I got the idea of this concept. Like it!